powernap/man/powerwake.1
.TH powerwake 1 "26 Jun 2009" powerwake "powerwake"
.SH NAME
powerwake \- smart utilty for remotely waking sleeping systems

.SH SYNOPSIS
.BI "powerwake [\-b|\-\-broadcast BROADCAST_IP] [\-m|\-\-method METHOD] TARGET_MAC|TARGET_IP|TARGET_HOST"

.SH OPTIONS
.TP
.B \-b|\-\-broadcast BROADCAST_IP
Optional parameter specifying the broadcast ip.  If unspecified, the default is 255.255.255.255.
.TP
.B \-m|\-\-method METHOD
Currently, 'wol' is the only supported method.
.TP
.B TARGET_MAC|TARGET_IP|TARGET_HOST
The target system to powerwake may be identified by MAC address, or by an IP/hostname cached in (the current \fBarp\fP(8) table, \fI/etc/ethers\fP, \fI/var/cache/powerwake/ethers\fP).

.SH DESCRIPTION
The powerwake utility compliments the powernap(8) daemon.  Administrators can run the powernap utility on remote hosts, allowing them to sleep when un-utilitized.  To awake the remote system, use powerwake.

Powerwake is designed to be "smart", and support multiple methods of waking remote hosts, such as wakeonlan, ipmi, and others.  Currently, wakeonlan is the only implemented mechanism, though others are intended.

Powerwake is also "smart" in that it can take MAC address, hostnames, or ip addresses.  In the case of a hostname or ip address, powerwake will first load a dynamically maintained cache file \fI/var/cache/powerwake/ethers\fP, then a statically administered configuration file \fI/etc/ethers\fP, and finally use the arp table to determine the target MAC address.  The cache will be updated accordingly.

.TP
\fIhttp://launchpad.net/powernap\fP
.PD

.SH FILES
\fI/etc/ethers\fP, \fI/var/cache/powerwake/ethers\fP

.SH SEE ALSO
\fBarp\fP(8)

.SH AUTHOR
This manpage and the utility was written by Dustin Kirkland <kirkland@canonical.com> for Ubuntu systems (but may be used by others).  Permission is granted to copy, distribute and/or modify this document under the terms of the GNU General Public License, Version 3 published by the Free Software Foundation.

On Debian systems, the complete text of the GNU General Public License can be found in /usr/share/common-licenses/GPL.


powernap/ChangeLog
powernap (2.18) released; urgency=low

  [ Frank Steinberg ]
  * LoadMonitor: handle float values correctly (LP: #1070696)
  * TCPMonitor: handle IPv6 connections correctly (LP: #1070695)

  [ Andres Rodriguez ]
  * debian/powernap.default: Add so that powernap does not start by default.

 -- Andres Rodriguez <andreserl@ubuntu.com>  Mon, 04 Jun 2012 12:59:15 -0400

powernap (2.17-0ubuntu1) precise; urgency=low

  [ Bryce Harrington ]
  * actions/{service,kernel_module}: Add scripts to handle services and
    kernel modules on powersave mode.

  [ Andres Rodriguez ]
  * Re-organize source to differ between pm-utils actions and powernap
    actions.
    - debian/powernap-common.install: Update.
    - powernap.spec: Update.
  * Disable SERVICES and KERN_MODULES on PowerSave mode (LP: #887684):
    - powernap/powernap.py: Load from config.
    - debian/powernap-common.install: Install new scripts.
    - Add take powersave action and its recover equivalent.
  * actions/pm/01cpu_online: Disable by default.

 -- Andres Rodriguez <andreserl@ubuntu.com>  Thu, 01 Dec 2011 12:43:51 -0500

powernap (2.16-0ubuntu1) precise; urgency=low

  [ Andres Rodriguez ]
  * debian/control: Move python-scapy Depends to Suggests.
  * Do not use awk to parse output. (LP: #879591)
    - bin/powerwake
    - powerwake/powerwake.py

  [ Eric Sandeen ]
  * powernap.spec: cleaned up a bit for better building/installation on RH
    systems

  [ Remi Druilhe ]
  * Dynamically reload configuration file (LP: #887680)

 -- Andres Rodriguez <andreserl@ubuntu.com>  Tue, 29 Nov 2011 14:42:41 -0500

powernap (2.15-0ubuntu1) oneiric; urgency=low

  * actions/eth_speed: Disable by default to not experience connectivity
    loss on certain network configs. (LP: #730210)

 -- Andres Rodriguez <andreserl@ubuntu.com>  Fri, 09 Sep 2011 15:32:30 -0400

powernap (2.14-0ubuntu1) oneiric; urgency=low

  * bin/powerwake:
    - fix the powerake mac caching methodology, which was somewhat broken

 -- Dustin Kirkland <kirkland@ubuntu.com>  Mon, 22 Aug 2011 00:38:17 -0500

powernap (2.13-0ubuntu1) oneiric; urgency=low

  [ Dustin Kirkland ]
  * sbin/powernapd:
    - make sure that powerwake-now works, even if a system has gone into
      powersave mode by other means (erhm, gnome-power-settings, erhm)
  * actions/01cpu_online:
    - drop the nasty sleep 10 hack around gnome-settings-daemon bugginess;
      doesn't seem to be a problem at this point in Oneiric;  If this bug
      shows back up, we need to solve it in a better way
  * actions/video:
    - drop incorrect inline docs
  * config:
    - now that the default is powersave mode, adjust the absent time and
      grace period; it takes fractions of a second to go in and out of
      powersave mode
  * actions/00flag, debian/powernap.upstart, sbin/powernapd: LP: #829618
    - use a flag in shared memory that notes the powersave status of the
      system
    - set/remove that flag in the 00flag script in the pm-utils suite
    - set/remove/query that same flag in powernapd itself
  * debian/control: LP: #637635
    - depend on anacron, to better handle server cronjobs on systems that
      might have been asleep
  * powernap/monitors/InputMonitor.py: LP: #826278
    - test device path existence before opening, fixes nasty crash

 -- Dustin Kirkland <kirkland@ubuntu.com>  Fri, 19 Aug 2011 09:57:07 -0700

powernap (2.12-0ubuntu1) oneiric; urgency=low

  [ Andres Rodriguez ]
  * Add Initial PowerNap (powerwaked) Server Capabilities:
    - powerwake/*: Add module and Initial ARPMonitor (LP: #737479).
    - powerwaked.conf: Add config file.
    - powerwaked: Add daemon.
    - powerwake-monitor: Add tool to add, delete, list, etc.
  * Package powernap-server:
    - debian/control: Add powerwaked, powerwake-common, powernap-server
    - debian/{powerwaked,powerwake-common}.install: Add
    - debian/powerwaked.upstart: Add upstart job.

  [ Dustin Kirkland ]
  * bin/powerwake:
    - only seed the arp cache if the argument is not already a mac address

 -- Andres Rodriguez <andreserl@ubuntu.com>  Tue, 09 Aug 2011 15:28:35 -0400

powernap (2.11-0ubuntu1) oneiric; urgency=low

  * actions/01cpu_online: Delay disabling of CPU's when machine is Desktop
    to allow it to apply template and not crash.

 -- Andres Rodriguez <andreserl@ubuntu.com>  Mon, 20 Jun 2011 18:08:09 -0400

powernap (2.10-0ubuntu1) oneiric; urgency=low

  * Add support for port ranges to TCPMonitor (LP: #760729)
    - powernap/monitors/TCPMonitor.py: Add support.
    - powernap/powernap.py: Ensure that ports are now passed as strings.
    - config: Update to reflect new feature.
  * debian/control:
    - Bump Build-Depends on debhelper.
    - Change python supported version.
    - Update Bzr-VCS to current.
    - Set Dustin as Original maintainer and set myself as maintainer.

 -- Andres Rodriguez <andreserl@ubuntu.com>  Thu, 09 Jun 2011 16:42:59 -0400

powernap (2.9-0ubuntu1) oneiric; urgency=low

  * sbin/powernap, sbin/powernapd:
    - add an indicator to show when powernap is engaged, remove after waking
  * === added directory debian/source, debian/source/format:
    - add this back in (got a fix for release scripts)

 -- Dustin Kirkland <kirkland@ubuntu.com>  Thu, 02 Jun 2011 17:43:23 -0400

powernap (2.8-0ubuntu1) oneiric; urgency=low

  [ Andres Rodriguez ]
  * Switch to dpkg-source 3.0 (quilt) format
  * Switch to dh_python2 instead of dh_pycentral
  * sbin/powernap: ACTION_METHOD is now passed as an argument.
  * Second Stage Action Method (LP: #711521)
    - config: Add config options.
    - powernap/powernap.py: Add loading of config values.
    - sbin/powernapd: Add logic to handle second stage action.

  [ Dustin Kirkland ]
  * config: remove trailing whitespace
  * debian/source/format, === removed directory debian/source:
    - drop these for now, as they're breaking the release scripts

 -- Dustin Kirkland <kirkland@ubuntu.com>  Thu, 02 Jun 2011 16:51:20 -0400

powernap (2.7-0ubuntu1) oneiric; urgency=low


powernap/powerwake/powerwake.py
#!/usr/bin/python
#
#    powerwake.py - handles powerwaked config and initializes Monitors.
#
#    Copyright (C) 2011 Canonical Ltd.
#
#    Authors: Andres Rodriguez <andreserl@canonical.com>
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, version 3 of the License.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.

import ConfigParser, sys, re, os
#from monitors import ARPMonitor 

class PowerWake:

    def __init__(self):
        self.PKG = "powerwake"
        self.CONFIG = "/etc/powernap/powerwaked.conf"
        self.ACTION = "/usr/bin/powerwake"
        self.INTERVAL_SECONDS = int(1)
        self.DEBUG = int(0)
        self.MONITORS = []
        # Load default config file (/etc/powernap/config)
        self.load_config_file()

    def load_config_file(self):
        cfg = ConfigParser.ConfigParser()
        cfg.read(self.CONFIG)

        try:
            # Load items in DEFAULT section
            defaults = cfg.items(self.PKG)
            for items in defaults:
                self.set_default_values(items[0], items[1])

            monitors_config = cfg.sections()
            for monitor in monitors_config:
                for items in cfg.items(monitor):
                    self.load_monitors_config(monitor, items)
        except:
            pass

    def set_default_values(self, var, value):
        if var == "interval_seconds":
            self.INTERVAL_SECONDS = eval(value)
        if var == "debug":
            self.DEBUG = eval(value)
        if var == "action":
            self.ACTION = eval(value)
        if var == "warn":
            if value == "y" or value == "yes":
                self.WARN = True

    def load_monitors_config(self, monitor, items):
        if monitor == "ARPMonitor" and (items[1] == "y" or items[1] == "yes"):
            self.MONITORS.append({"monitor":monitor, "cache":self.get_monitored_hosts(monitor.lower())})

    def get_monitors(self):
        from monitors import ARPMonitor 
        monitor = []
        for config in self.MONITORS:
            if config["monitor"] == "ARPMonitor":
                p = ARPMonitor.ARPMonitor(config)
            monitor.append(p)

        return monitor

    def get_monitored_hosts(self, monitor):
        host_to_mac = {}
        for file in ["/etc/powernap/powerwaked.%s.ethers" % monitor]:
            if os.path.exists(file):
                f = open(file, 'r')
                for i in f.readlines():
                    try:
                        (m, h) = i.split()
                        host_to_mac[h] = m
                    except:
                        pass
                f.close()
        return host_to_mac

    def set_monitored_hosts(self, host_to_mac, monitor):
        path = ["/etc/powernap/powerwaked.%s.ethers" % monitor]
        for file in path:
            if not os.path.exists(file):
                f = open(file, 'a')
                f.close()
        for file in ["/etc/powernap/powerwaked.%s.ethers" % monitor]:
            if os.access(file, os.W_OK):
                f = open(file, 'w')
                for h in host_to_mac:
                    if self.is_mac(host_to_mac[h]):
                        f.write("%s %s\n" % (host_to_mac[h], h))
                f.close()

    def get_mac_or_ip_from_arp(self, host):
        mac_or_ip = None
        for i in os.popen("/usr/sbin/arp -n"):
            m = i.split()[2]
            h = i.split()[0]
            if self.is_mac(host) and host == m and self.is_ip(h):
                mac_or_ip = h
                break
            if self.is_ip(host) and host == h and self.is_mac(m):
                mac_or_ip = m
                break
        #if not mac_or_ip:
        #    raise BaseException("Error")
        return mac_or_ip

    def is_ip(self, ip):
        r1 = re.compile('^[0-9]+\.[0-9]+\.[0-9]+\.[0-9]+$')
        if r1.match(ip):
            return True
        else:
            return False

    def is_mac(self, mac):
        r1 = re.compile('^[0-9a-fA-F]{12}$')
        r2 = re.compile('^[0-9a-fA-F]{2}.[0-9a-fA-F]{2}.[0-9a-fA-F]{2}.[0-9a-fA-F]{2}.[0-9a-fA-F]{2}.[0-9a-fA-F]{2}$')
        if r1.match(mac) or r2.match(mac):
            return 1
        else:
            return 0

#### --------------------------------- from powerwake --------------------------------####

    # Source the cached, known arp entries
    def get_arp_cache(self):
        host_to_mac = {}
        #for file in ["/var/cache/%s/ethers" % self.PKG, "/etc/ethers", "%s/.cache/ethers" % HOME]:
        for file in ["/var/cache/%s/ethers" % self.PKG, "/etc/ethers"]:
            if os.path.exists(file):
                f = open(file, 'r')
                for i in f.readlines():
                    try:
                        (m, h) = i.split()
                        host_to_mac[h] = m
                    except:
                        pass
                f.close()
        return host_to_mac

    # Source the current, working arp table
    def get_arp_current(self, host_to_mac):
        # Load hostnames
        for i in os.popen("/usr/sbin/arp"):
            m = i.split()[2]
            h = i.split()[0]
            if is_mac(m):
                host_to_mac[h] = m
        # Load ip addresses
        for i in os.popen("/usr/sbin/arp -n"):
            m = i.split()[2]
            h = i.split()[0]
            if is_mac(m):
                host_to_mac[h] = m
        return host_to_mac

    def write_arp_cache(self, host_to_mac):
        if not os.access("%s/.cache/" % HOME, os.W_OK):
            return
        if not os.path.exists("%s/.cache/" % HOME):
            os.makedirs("%s/.cache/" % HOME)
        f = open("%s/.cache/ethers" % HOME, 'a')
        f.close()
        for file in ["/var/cache/%s/ethers" % PKG, "%s/.cache/ethers" % HOME]:
            if os.access(file, os.W_OK):
                f = open(file, 'w')
                for h in host_to_mac:
                    if self.is_mac(host_to_mac[h]):
                        f.write("%s %s\n" % (host_to_mac[h], h))
                f.close()
